File: app/src/main/java/com/jzoft/ygohelper/biz/DuelCalculator.java
package com.jzoft.ygohelper.biz;

import java.util.LinkedList;
import java.util.NoSuchElementException;

/**
 * Created by jjimenez on 9/03/17.
 */
public class DuelCalculator {
    private static final int SIGN_SWITCH = -1;

    private final LinkedList<Damage> log;
    private int pre;
    private int sign;
    private boolean half;
    private boolean quit;
    private boolean hold;

    public DuelCalculator() {
        pre = 0;
        reset();
        log = new LinkedList<Damage>();
    }

    private void reset() {
        sign = -1;
        hold = false;
        resetStatus();
    }

    private void resetStatus() {
        half = false;
        quit = false;
    }

    public String getPre() {
        if (quit) return "QUIT";
        if (half) return "1/2";
        return "" + absolut(pre);
    }

    public void undo() {
        try {
            Damage last = log.getLast();
            pre = absolut(last.amount);
            log.removeLast();
        } catch (NoSuchElementException e) {
        }
    }

    private int absolut(int amount) {
        return amount < 0 ? -amount : amount;
    }

    public void half() {
        half = true;
    }

    public void quit() {
        quit = true;
    }

    public void toggleSign() {
        sign *= SIGN_SWITCH;
    }

    public boolean isNegative() {
        return sign < 0;
    }

    public void addNumber(int number) {
        resetStatus();
        if (hold) {
            pre = (pre * 10) + number;
        } else {
            hold = true;
            pre = number;
        }
    }

    public void mark(Player player) {
        int amount = getAmountByStatus(player);
        if (amount != 0 && validGame()) {
            log.addLast(new Damage(player, amount));
            pre = absolut(amount);
            reset();
        }
    }

    public boolean validGame() {
        return getLife(Player.A) > 0 && getLife(Player.B) > 0;
    }

    private int getAmountByStatus(Player player) {
        if (half) return -(getLife(player) / 2);
        if (quit) return -getLife(player);
        return sign * getAmountByPre();
    }

    private int getAmountByPre() {
        if (pre == 50) return 50;
        if (pre < 100) return pre * 100;
        return pre;
    }

    public int getLife(Player player) {
        int total = 8000 + totalAmountOf(player);
        return total < 0 ? 0 : total;
    }

    private int totalAmountOf(Player player) {
        int total = 0;
        for (Damage d : log)
            if (d.player == player) total += d.amount;
        return total;
    }

    public String displayLog() {
        StringBuilder builder = new StringBuilder(" 8000| 8000\n");
        int a = 8000;
        int b = 8000;
        for (Damage d : log) {
            int amount = d.amount;
            if (d.player == Player.A) {
                a += amount;
                builder.append(amount > 0 ? '+' : '-');
                appendAmount(builder, amount);
                builder.append("|     ");
            } else {
                b += amount;
                builder.append("     |");
                builder.append(amount > 0 ? '+' : '-');
                appendAmount(builder, amount);
            }
            appendTotal(builder, a, b);
        }
        return builder.toString();
    }

    private void appendTotal(StringBuilder builder, int a, int b) {
        builder.append('\n');
        builder.append(' ');
        appendAmount(builder, a < 0 ? 0 : a);
        builder.append("|");
        builder.append(' ');
        appendAmount(builder, b < 0 ? 0 : b);
        builder.append('\n');
    }

    private void appendAmount(StringBuilder builder, int amount) {
        int absolut = absolut(amount);
        if (absolut < 1000) builder.append(' ');
        if (absolut < 100) builder.append(' ');
        builder.append(absolut);
    }


    public enum Player {A, B;}

    private static class Damage {
        Player player;
        int amount;

        public Damage(Player player, int amount) {
            this.player = player;
            this.amount = amount;
        }
    }
}

File: app/src/main/java/com/jzoft/ygohelper/biz/DuelCalculator.kt
package com.jzoft.ygohelper.biz

import java.util.LinkedList
import java.util.NoSuchElementException

/**
 * Created by jjimenez on 9/03/17.
 */
class DuelCalculator {

    private val log: LinkedList<Damage>
    private var pre: Int = 0
    private var sign: Int = 0
    private var half: Boolean = false
    private var quit: Boolean = false
    private var hold: Boolean = false

    val isNegative: Boolean
        get() = sign < 0

    private val amountByPre: Int
        get() {
            if (pre == 50) return 50
            return if (pre < 100) pre * 100 else pre
        }

    init {
        pre = 0
        reset()
        log = LinkedList()
    }

    private fun reset() {
        sign = -1
        hold = false
        resetStatus()
    }

    private fun resetStatus() {
        half = false
        quit = false
    }

    fun preview(): String {
        if (quit) return "QUIT"
        return if (half) "1/2" else "" + absolut(pre)
    }

    fun undo() {
        try {
            val last = log.last
            pre = absolut(last.amount)
            log.removeLast()
        } catch (e: NoSuchElementException) {
        }

    }

    private fun absolut(amount: Int): Int {
        return if (amount < 0) -amount else amount
    }

    fun half() {
        half = true
    }

    fun quit() {
        quit = true
    }

    fun toggleSign() {
        sign *= SIGN_SWITCH
    }

    fun addNumber(number: Int) {
        resetStatus()
        if (hold) {
            pre = pre * 10 + number
        } else {
            hold = true
            pre = number
        }
    }

    fun mark(player: Player) {
        val amount = getAmountByStatus(player)
        if (amount != 0 && validGame()) {
            log.addLast(Damage(player, amount))
            pre = absolut(amount)
            reset()
        }
    }

    fun validGame(): Boolean {
        return getLife(Player.A) > 0 && getLife(Player.B) > 0
    }

    private fun getAmountByStatus(player: Player): Int {
        if (half) return -(getLife(player) / 2)
        return if (quit) -getLife(player) else sign * amountByPre
    }

    fun getLife(player: Player): Int {
        val total = 8000 + totalAmountOf(player)
        return if (total < 0) 0 else total
    }

    private fun totalAmountOf(player: Player): Int {
        var total = 0
        for (d in log)
            if (d.player == player) total += d.amount
        return total
    }

    fun displayLog(): String {
        val builder = StringBuilder(" 8000| 8000\n")
        var a = 8000
        var b = 8000
        for (d in log) {
            val amount = d.amount
            if (d.player == Player.A) {
                a += amount
                builder.append(if (amount > 0) '+' else '-')
                appendAmount(builder, amount)
                builder.append("|     ")
            } else {
                b += amount
                builder.append("     |")
                builder.append(if (amount > 0) '+' else '-')
                appendAmount(builder, amount)
            }
            appendTotal(builder, a, b)
        }
        return builder.toString()
    }

    private fun appendTotal(builder: StringBuilder, a: Int, b: Int) {
        builder.append('\n')
        builder.append(' ')
        appendAmount(builder, if (a < 0) 0 else a)
        builder.append("|")
        builder.append(' ')
        appendAmount(builder, if (b < 0) 0 else b)
        builder.append('\n')
    }

    private fun appendAmount(builder: StringBuilder, amount: Int) {
        val absolut = absolut(amount)
        if (absolut < 1000) builder.append(' ')
        if (absolut < 100) builder.append(' ')
        builder.append(absolut)
    }


    enum class Player {
        A, B
    }

    private class Damage(internal var player: Player, internal var amount: Int)

    companion object {
        private val SIGN_SWITCH = -1
    }
}
